File: glyzier-backend/src/main/java/com/glyzier/service/UserService.java
package com.glyzier.service;

import com.glyzier.dto.UpdateProfileRequest;
import com.glyzier.dto.ChangePasswordRequest;
import com.glyzier.model.Users;
import com.glyzier.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

/**
 * User Service
 * 
 * This service handles user-related operations beyond authentication.
 * It provides methods to retrieve user information, particularly for
 * authenticated users accessing their own data.
 * 
 * Key responsibilities:
 * - Get the currently authenticated user
 * - Retrieve user details by ID or email
 * - Manage user profile information
 * 
 * @author Glyzier Team
 * @version 1.0
 */
@Service
public class UserService {

    /**
     * Repository for database operations on Users
     */
    @Autowired
    private UserRepository userRepository;

    /**
     * Password encoder for hashing passwords
     * Used for password change operations
     */
    @Autowired
    private PasswordEncoder passwordEncoder;

    /**
     * Get the currently authenticated user
     * 
     * This method:
     * 1. Retrieves the authentication object from Spring Security's SecurityContext
     * 2. Extracts the username (email) from the authenticated principal
     * 3. Looks up and returns the full Users entity from the database
     * 
     * The SecurityContext is set by our JwtAuthFilter after validating the JWT token.
     * 
     * @return Users entity of the currently authenticated user
     * @throws UsernameNotFoundException if the authenticated user is not found in database
     * @throws RuntimeException if no user is authenticated
     */
    public Users getCurrentUser() {
        // Get the current authentication from the security context
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();

        // Check if a user is authenticated
        if (authentication == null || !authentication.isAuthenticated()) {
            throw new RuntimeException("No authenticated user found");
        }

        // Get the principal (UserDetails) from the authentication
        Object principal = authentication.getPrincipal();
        
        // Extract the username (email)
        String email;
        if (principal instanceof UserDetails) {
            email = ((UserDetails) principal).getUsername();
        } else {
            email = principal.toString();
        }

        // Look up the user in the database and return
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found: " + email));
    }

    /**
     * Get a user by their ID
     * 
     * @param userid The user's unique identifier
     * @return Users entity
     * @throws RuntimeException if user is not found
     */
    public Users getUserById(Long userid) {
        return userRepository.findById(userid)
                .orElseThrow(() -> new RuntimeException("User not found with ID: " + userid));
    }

    /**
     * Get a user by their email
     * 
     * @param email The user's email address
     * @return Users entity
     * @throws UsernameNotFoundException if user is not found
     */
    public Users getUserByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new UsernameNotFoundException("User not found with email: " + email));
    }

    /**
     * Check if the current user is a seller
     * 
     * @return true if the current user has a seller account, false otherwise
     */
    public boolean isCurrentUserSeller() {
        Users currentUser = getCurrentUser();
        return currentUser.getSeller() != null;
    }

    /**
     * Update user profile information
     * 
     * This method allows users to update their display name and phone number.
     * Email cannot be changed through this method for security reasons.
     * 
     * Module 14 Implementation:
     * - Updates display name (required)
     * - Updates phone number (optional, can be null or empty)
     * - Saves changes to database
     * - Returns updated user entity
     * 
     * @param request UpdateProfileRequest containing new display name and phone number
     * @return Updated Users entity
     * @throws RuntimeException if user is not found
     */
    @Transactional
    public Users updateProfile(UpdateProfileRequest request) {
        // Get the current authenticated user
        Users currentUser = getCurrentUser();

        // Update the fields from the request
        currentUser.setDisplayname(request.getDisplayname());
        
        // Phone number is optional - set even if null or empty
        currentUser.setPhonenumber(request.getPhonenumber());

        // Save and return the updated user
        return userRepository.save(currentUser);
    }

    /**
     * Change user password
     * 
     * This method allows users to change their password securely.
     * It requires verification of the current password before allowing the change.
     * 
     * Security measures:
     * 1. Verifies current password matches stored hash
     * 2. Validates new password matches confirmation
     * 3. Encrypts new password before storage
     * 
     * Module 14 Implementation:
     * - Validates current password is correct
     * - Checks new password and confirmation match
     * - Hashes new password using BCrypt
     * - Updates password in database
     * 
     * @param request ChangePasswordRequest containing current and new passwords
     * @throws RuntimeException if current password is incorrect
     * @throws RuntimeException if new password and confirmation don't match
     * @throws RuntimeException if user is not found
     */
    @Transactional
    public void changePassword(ChangePasswordRequest request) {
        // Get the current authenticated user
        Users currentUser = getCurrentUser();

        // Verify current password is correct
        // passwordEncoder.matches() compares plain text with BCrypt hash
        if (!passwordEncoder.matches(request.getCurrentPassword(), currentUser.getPassword())) {
            throw new RuntimeException("Current password is incorrect");
        }

        // Verify new password and confirmation match
        if (!request.getNewPassword().equals(request.getConfirmPassword())) {
            throw new RuntimeException("New password and confirmation do not match");
        }

        // Encrypt the new password and update
        currentUser.setPassword(passwordEncoder.encode(request.getNewPassword()));

        // Save the updated user
        userRepository.save(currentUser);
    }
}

File: glyzier-frontend/src/services/userService.js
/**
 * User Service - Frontend API calls for user profile management
 * 
 * This service provides functions for user-related operations:
 * - Get current user information
 * - Update user profile (display name, phone number)
 * - Change password
 * 
 * All functions use the pre-configured api instance which automatically
 * adds the JWT token to requests.
 * 
 * Module 14 - Basic User Profile implementation
 * 
 * @author Glyzier Team
 * @version 1.0
 */

import api from './api';

/**
 * Get current authenticated user's information
 * 
 * Fetches the profile data of the currently logged-in user.
 * JWT token is automatically added from localStorage by api interceptor.
 * 
 * Endpoint: GET /api/users/me
 * 
 * @returns {Promise<Object>} User data including:
 *   - userid: User's unique ID
 *   - email: User's email address
 *   - displayname: User's display name
 *   - phonenumber: User's phone number (optional)
 *   - isSeller: Whether user is a seller
 *   - createdAt: Account creation timestamp
 * 
 * @throws {Error} If request fails (network error, unauthorized, etc.)
 * 
 * @example
 * const user = await getCurrentUser();
 * console.log(user.displayname); // "John Doe"
 */
export const getCurrentUser = async () => {
  const response = await api.get('/api/users/me');
  return response.data;
};

/**
 * Update user profile information
 * 
 * Updates the current user's display name and/or phone number.
 * Email cannot be changed through this endpoint.
 * 
 * Endpoint: PUT /api/users/profile
 * 
 * @param {Object} profileData - Profile data to update
 * @param {string} profileData.displayname - New display name (required, 2-100 chars)
 * @param {string} [profileData.phonenumber] - New phone number (optional, 10-20 chars)
 * 
 * @returns {Promise<Object>} Response containing:
 *   - message: Success message
 *   - user: Updated user object
 * 
 * @throws {Error} If validation fails or request fails
 * 
 * @example
 * const result = await updateProfile({
 *   displayname: "Jane Doe",
 *   phonenumber: "+1 234 567 8900"
 * });
 * console.log(result.message); // "Profile updated successfully"
 */
export const updateProfile = async (profileData) => {
  const response = await api.put('/api/users/profile', profileData);
  return response.data;
};

/**
 * Change user password
 * 
 * Changes the current user's password. Requires verification of the
 * current password before allowing the change. New password must match
 * confirmation.
 * 
 * Endpoint: PUT /api/users/change-password
 * 
 * @param {Object} passwordData - Password change data
 * @param {string} passwordData.currentPassword - Current password for verification
 * @param {string} passwordData.newPassword - New password (min 8 chars)
 * @param {string} passwordData.confirmPassword - Confirmation of new password
 * 
 * @returns {Promise<Object>} Response containing:
 *   - message: Success message
 * 
 * @throws {Error} If current password is incorrect, passwords don't match, or validation fails
 * 
 * @example
 * await changePassword({
 *   currentPassword: "oldpass123",
 *   newPassword: "newpass456",
 *   confirmPassword: "newpass456"
 * });
 * // Password changed successfully
 */
export const changePassword = async (passwordData) => {
  const response = await api.put('/api/users/change-password', passwordData);
  return response.data;
};

// Export as default object for convenience
export default {
  getCurrentUser,
  updateProfile,
  changePassword
};
